File: ios-sdk/Sources/MxLSdk/Crash/CrashReporter.swift
import Foundation
import UIKit

/**
 * Crash reporter for capturing and reporting crashes.
 */
internal class CrashReporter {
    private static var isInitialized = false
    private static var originalExceptionHandler: NSUncaughtExceptionHandler?
    private static var signalHandlers: [Int32: sig_t] = [:]
    
    /**
     * Initialize crash reporting.
     */
    static func initialize(configuration: SdkConfiguration) {
        guard configuration.enableCrashReporting, !isInitialized else {
            return
        }
        
        // Set up exception handler for Objective-C exceptions
        originalExceptionHandler = NSGetUncaughtExceptionHandler()
        NSSetUncaughtExceptionHandler { exception in
            captureException(exception)
            originalExceptionHandler?(exception)
        }
        
        // Set up signal handlers for C/C++ crashes
        setupSignalHandlers()
        
        isInitialized = true
        Logger.shared.info("Crash reporting initialized")
    }
    
    /**
     * Set up signal handlers for native crashes.
     */
    private static func setupSignalHandlers() {
        let signals: [Int32] = [SIGABRT, SIGBUS, SIGFPE, SIGILL, SIGSEGV, SIGTRAP]
        
        for signal in signals {
            signalHandlers[signal] = signal(signal, { signal in
                captureSignal(signal)
                // Re-raise signal with default handler
                signal(signal, SIG_DFL)
                raise(signal)
            })
        }
    }
    
    /**
     * Capture Objective-C exception.
     */
    private static func captureException(_ exception: NSException) {
        let sdkState = MxLSdk.getSdkState()
        guard let sdkState = sdkState else { return }
        
        let stackTrace = exception.callStackSymbols.joined(separator: "\n")
        let threadInfo = getThreadInfo()
        
        let crashEvent = CrashEvent(
            sessionId: sdkState.sessionManager.getSessionId() ?? "unknown",
            userId: sdkState.sessionManager.getUserId(),
            exceptionType: exception.name.rawValue,
            exceptionMessage: exception.reason ?? "",
            stackTrace: stackTrace,
            threadInfo: threadInfo,
            timestamp: Date().timeIntervalSince1970 * 1000,
            deviceInfo: getDeviceInfo()
        )
        
        storeCrash(crashEvent, sdkState: sdkState)
    }
    
    /**
     * Capture signal-based crash.
     */
    private static func captureSignal(_ signal: Int32) {
        let sdkState = MxLSdk.getSdkState()
        guard let sdkState = sdkState else { return }
        
        let signalName = getSignalName(signal)
        let stackTrace = Thread.callStackSymbols.joined(separator: "\n")
        let threadInfo = getThreadInfo()
        
        let crashEvent = CrashEvent(
            sessionId: sdkState.sessionManager.getSessionId() ?? "unknown",
            userId: sdkState.sessionManager.getUserId(),
            exceptionType: "Signal: \(signalName)",
            exceptionMessage: "Application crashed with signal \(signalName)",
            stackTrace: stackTrace,
            threadInfo: threadInfo,
            timestamp: Date().timeIntervalSince1970 * 1000,
            deviceInfo: getDeviceInfo()
        )
        
        storeCrash(crashEvent, sdkState: sdkState)
    }
    
    /**
     * Get signal name.
     */
    private static func getSignalName(_ signal: Int32) -> String {
        switch signal {
        case SIGABRT: return "SIGABRT"
        case SIGBUS: return "SIGBUS"
        case SIGFPE: return "SIGFPE"
        case SIGILL: return "SIGILL"
        case SIGSEGV: return "SIGSEGV"
        case SIGTRAP: return "SIGTRAP"
        default: return "UNKNOWN"
        }
    }
    
    /**
     * Get thread information.
     */
    private static func getThreadInfo() -> ThreadInfo {
        let thread = Thread.current
        return ThreadInfo(
            name: thread.name.isEmpty ? "Main" : thread.name,
            isMainThread: Thread.isMainThread,
            threadNumber: thread.hashValue
        )
    }
    
    /**
     * Get device information.
     */
    private static func getDeviceInfo() -> DeviceInfo {
        let device = UIDevice.current
        let systemVersion = device.systemVersion
        let model = device.model
        let name = device.name
        
        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown"
        let buildNumber = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "unknown"
        
        return DeviceInfo(
            platform: "iOS",
            osVersion: systemVersion,
            deviceModel: model,
            deviceName: name,
            appVersion: appVersion,
            buildNumber: buildNumber
        )
    }
    
    /**
     * Store crash event.
     */
    private static func storeCrash(_ crashEvent: CrashEvent, sdkState: SdkState) {
        // TODO: Store in database and queue for upload
        Logger.shared.error("Crash captured: \(crashEvent.exceptionType)")
    }
    
    /**
     * Report a custom error manually.
     */
    static func reportError(_ error: Error, context: [String: Any] = [:]) {
        let sdkState = MxLSdk.getSdkState()
        guard let sdkState = sdkState else { return }
        
        let stackTrace = Thread.callStackSymbols.joined(separator: "\n")
        let threadInfo = getThreadInfo()
        
        let crashEvent = CrashEvent(
            sessionId: sdkState.sessionManager.getSessionId() ?? "unknown",
            userId: sdkState.sessionManager.getUserId(),
            exceptionType: String(describing: type(of: error)),
            exceptionMessage: error.localizedDescription,
            stackTrace: stackTrace,
            threadInfo: threadInfo,
            timestamp: Date().timeIntervalSince1970 * 1000,
            deviceInfo: getDeviceInfo(),
            customContext: context
        )
        
        storeCrash(crashEvent, sdkState: sdkState)
    }
}

struct CrashEvent {
    let sessionId: String
    let userId: String?
    let exceptionType: String
    let exceptionMessage: String
    let stackTrace: String
    let threadInfo: ThreadInfo
    let timestamp: TimeInterval
    let deviceInfo: DeviceInfo
    let customContext: [String: Any]
}

struct ThreadInfo {
    let name: String
    let isMainThread: Bool
    let threadNumber: Int
}

struct DeviceInfo {
    let platform: String
    let osVersion: String
    let deviceModel: String
    let deviceName: String
    let appVersion: String
    let buildNumber: String
}

File: android-sdk/sdk/src/main/java/com/adlcom/mxl/sdk/crash/CrashReporter.kt
package com.adlcom.mxl.sdk.crash

import android.content.Context
import com.adlcom.mxl.sdk.config.SdkConfiguration
import com.adlcom.mxl.sdk.internal.SdkState
import com.adlcom.mxl.sdk.internal.getSdkState
import com.adlcom.mxl.sdk.logging.Logger
import com.adlcom.mxl.sdk.session.SessionManager
import java.io.PrintWriter
import java.io.StringWriter

/**
 * Crash reporter for capturing and reporting crashes.
 */
object CrashReporter {
    private var originalHandler: Thread.UncaughtExceptionHandler? = null
    private var isInitialized = false

    /**
     * Initialize crash reporting.
     */
    fun initialize(context: Context, configuration: SdkConfiguration) {
        if (!configuration.enableCrashReporting || isInitialized) {
            return
        }

        originalHandler = Thread.getDefaultUncaughtExceptionHandler()

        val crashHandler = object : Thread.UncaughtExceptionHandler {
            override fun uncaughtException(thread: Thread, throwable: Throwable) {
                try {
                    captureCrash(thread, throwable)
                } catch (e: Exception) {
                    Logger.e("Error capturing crash", e)
                } finally {
                    // Call original handler
                    originalHandler?.uncaughtException(thread, throwable)
                }
            }
        }

        Thread.setDefaultUncaughtExceptionHandler(crashHandler)
        isInitialized = true
        Logger.d("Crash reporting initialized")
    }

    /**
     * Capture a crash event.
     */
    private fun captureCrash(thread: Thread, throwable: Throwable) {
        val sdkState = getSdkState() ?: return

        val stackTrace = getStackTrace(throwable)
        val threadInfo = getThreadInfo(thread)

        val crashEvent = CrashEvent(
            sessionId = sdkState.sessionManager.getSessionId() ?: "unknown",
            userId = sdkState.sessionManager.getUserId(),
            exceptionType = throwable.javaClass.name,
            exceptionMessage = throwable.message ?: "",
            stackTrace = stackTrace,
            threadInfo = threadInfo,
            timestamp = System.currentTimeMillis(),
            deviceInfo = getDeviceInfo(sdkState.context)
        )

        // Store crash for later upload
        storeCrash(crashEvent, sdkState)
    }

    /**
     * Get stack trace as string.
     */
    private fun getStackTrace(throwable: Throwable): String {
        val sw = StringWriter()
        val pw = PrintWriter(sw)
        throwable.printStackTrace(pw)
        return sw.toString()
    }

    /**
     * Get thread information.
     */
    private fun getThreadInfo(thread: Thread): ThreadInfo {
        return ThreadInfo(
            name = thread.name,
            id = thread.id,
            priority = thread.priority,
            state = thread.state.name
        )
    }

    /**
     * Get device information.
     */
    private fun getDeviceInfo(context: Context): DeviceInfo {
        val packageManager = context.packageManager
        val packageInfo = packageManager.getPackageInfo(context.packageName, 0)

        return DeviceInfo(
            platform = "Android",
            osVersion = android.os.Build.VERSION.RELEASE,
            sdkVersion = android.os.Build.VERSION.SDK_INT,
            deviceModel = android.os.Build.MODEL,
            manufacturer = android.os.Build.MANUFACTURER,
            appVersion = packageInfo.versionName ?: "unknown",
            appVersionCode = packageInfo.longVersionCode
        )
    }

    /**
     * Store crash event for later upload.
     */
    private fun storeCrash(crashEvent: CrashEvent, sdkState: SdkState) {
        kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
            try {
                sdkState.storageManager.storeEvent(
                    eventType = "crash",
                    eventData = crashEvent,
                    sessionId = crashEvent.sessionId,
                    userId = crashEvent.userId
                )
                Logger.d("Crash event stored: ${crashEvent.exceptionType}")
            } catch (e: Exception) {
                Logger.e("Error storing crash event", e)
            }
        }
    }

    /**
     * Report a custom error manually.
     */
    fun reportError(throwable: Throwable, context: Map<String, Any> = emptyMap()) {
        val sdkState = getSdkState() ?: return

        val stackTrace = getStackTrace(throwable)
        val threadInfo = getThreadInfo(Thread.currentThread())

        val crashEvent = CrashEvent(
            sessionId = sdkState.sessionManager.getSessionId() ?: "unknown",
            userId = sdkState.sessionManager.getUserId(),
            exceptionType = throwable.javaClass.name,
            exceptionMessage = throwable.message ?: "",
            stackTrace = stackTrace,
            threadInfo = threadInfo,
            timestamp = System.currentTimeMillis(),
            deviceInfo = getDeviceInfo(sdkState.context),
            customContext = context
        )

        storeCrash(crashEvent, sdkState)
    }
}

data class CrashEvent(
    val sessionId: String,
    val userId: String?,
    val exceptionType: String,
    val exceptionMessage: String,
    val stackTrace: String,
    val threadInfo: ThreadInfo,
    val timestamp: Long,
    val deviceInfo: DeviceInfo,
    val customContext: Map<String, Any> = emptyMap()
)

data class ThreadInfo(
    val name: String,
    val id: Long,
    val priority: Int,
    val state: String
)

data class DeviceInfo(
    val platform: String,
    val osVersion: String,
    val sdkVersion: Int,
    val deviceModel: String,
    val manufacturer: String,
    val appVersion: String,
    val appVersionCode: Long
)

